gstd/src/common/reservations.rs

```diff
@@ -0,0 +1,328 @@
+    ///
+    /// `duration` is the block count within which the reservation must be used.
+    ///
+    /// # Underlying logics
+    ///
+    /// Executes for O(logN)..O(N), where N is a number of stored
+    /// reservations.
+    ///
+    /// All the reservations are kept sorted by amount in
+    /// ascending order(when amount is the same, they're sorted by time when
+    /// they expire) when inserted, so the closer inserted element to the
+    /// beginning of the underlying `Vec` the closer execution time will be
+    /// to O(N).
+    ///
+    /// Also, when the underlying `Vec` will allocate new memory
+    /// the attempt to clean expired reservations occurs to avoid memory
+    /// allocations.
+    pub fn reserve(&mut self, amount: u64, duration: u32) -> Result<()> {
+        let id = ReservationId::reserve(amount, duration)?;
+        let block_height = exec::block_height();
+        let valid_until = duration.saturating_add(block_height);
+
+        let insert_range_start = self
+            .0
+            .partition_point(|reservation| reservation.amount < amount);
+        let insert_range_end = self
+            .0
+            .partition_point(|reservation| reservation.amount <= amount);
+
+        let insert_to = self.0[insert_range_start..insert_range_end]
+            .binary_search_by(|reservation| reservation.valid_until.cmp(&valid_until));
+        let insert_to = if insert_range_start == self.0.len() {
+            self.0.len()
+        } else {
+            match insert_to {
+                Ok(pos) => pos + 1,
+                Err(pos) => pos,
+            }
+        };
+
+        // self.0 will allocate new memory.
+        if self.0.capacity() == self.0.len() {
+            self.cleanup(block_height);
+        }
+
+        self.0.insert(
+            insert_to,
+            Reservation {
+                id,
+                amount,
+                valid_until,
+            },
+        );
+
+        Ok(())
+    }
+
+    /// Find the appropriate reservation with reserved amount greater than or
+    /// equal to `amount`.
+    ///
+    /// If such a reservation is found, [`ReservationId`] is returned.
+    ///
+    /// # Underlying logics
+    ///
+    /// Executes for O(logN)..O(N), where N is amount of stored
+    /// reservations. When there's many expired reservations execution time
+    /// is closer to O(N).
+    ///
+    /// All the reservations are sorted by their amount and then by the time
+    /// when they'll expire (both are in ascending order) in underlying `Vec`,
+    /// so when one's trying to take reservation, reservation with the least
+    /// possible amount is found and if it's already expired, the search to the
+    /// end of underlying `Vec` occurs. After that, all the expired
+    /// reservations that were found in process of search are cleaned out.
+    ///
+    /// # See also
+    /// - [`ReservationId`] is used to reserve and unreserve gas amount
+    /// for program execution later.
```

Comment:
Seems this fn doesn’t pay attention on diff of expiration for two +- similar reservations

Reply:
Still I’d also return bn here or even Reservation structure itself